### PR TITLE
feat: implement `error-log-logger` plug-in

### LINKED ISSUE #2410: feat: plugin `error-log-logger`, collect error log and report to remote server by `cosocket`
all of the error log will be written into local files now.

```
$ ls logs
access.log  error.log
```

we can implement a new plugin which named `error-log-logger`, it'll collect the error log from API: https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/errlog.md

we can report those error log to `syslog-ng` server or HTTP server.

### PR DESCRIPTION
**Pull request by dabue**
### What this PR does / why we need it:

add a plugin to collect error log and report to remote TCP server .

fix #2410

you can set the plugin's config in config-default.yaml, likes below:
plugins:                          # plugin list
  - error-log-logger

and  update it links below:
step 1: update the attributes of the plugin

curl http://127.0.0.1:9080//apisix/admin/plugin_metadata/error-log-logger -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
  "host": "127.0.0.1",
  "port": 1999,
  "inactive_timeout": 1
}'

### CODE REVIEW

File: apisix/plugins/error-log-logger.lua
Comment (membphis) [inline apisix/plugins/error-log-logger.lua:118]: ditto

File: apisix/plugins/error-log-logger.lua
Comment (membphis) [inline apisix/plugins/error-log-logger.lua:128]: this way is wrong, please take a look at: https://github.com/apache/apisix/blob/master/apisix/plugins/http-logger.lua#L166

File: apisix/plugins/error-log-logger.lua
Comment (membphis) [inline apisix/plugins/error-log-logger.lua:168]: `for i,v in ipairs(table_name) do` this style is better for this case

File: apisix/plugins/error-log-logger.lua
Comment (membphis) [inline apisix/plugins/error-log-logger.lua:192]: ditto

File: apisix/plugins/error-log-logger.lua
Comment (membphis) [inline apisix/plugins/error-log-logger.lua:217]: remove this line

File: apisix/plugins/error-log-logger.lua
Comment (Firstsawyou) [inline apisix/plugins/error-log-logger.lua:68]: I think this `local _M = {}` format will look better, what do you think?

Comment (dabue) [reply apisix/plugins/error-log-logger.lua:68]: fixed it

**Comment by dabue** [comment]
@membphis  please review.

**Comment by membphis** [comment]
@dabue we can merge this PR after https://github.com/apache/apisix/pull/2596 merged.

I think this plugin depends on this PR.

**Comment by dabue** [comment]
@membphis  agree it. After  #2596 merged, I can use the backgroud timer to update this pr.

**Comment by dabue** [comment]
@membphis  this pr has been updated after  #2596  and used the new timer, please have a review.

**Comment by membphis** [comment]
> @membphis this pr has been updated after #2596 and used the new timer, please have a review.

got it, let me take a look ^_^

File: doc/plugins/error-log-logger.md
Comment (membphis) [inline doc/plugins/error-log-logger.md:70]:
```yaml
plugins:                          # plugin list
  ... ...
  - request-id
  - hmac-auth
  - api-breaker
  - error-log-logger              # enable plugin `error-log-logger`

File: doc/plugins/error-log-logger.md
Comment (membphis) [inline doc/plugins/error-log-logger.md:75]: add a example

File: doc/plugins/error-log-logger.md
Comment (membphis) [inline doc/plugins/error-log-logger.md:79]: `Step`

File: doc/plugins/error-log-logger.md
Comment (membphis) [inline doc/plugins/error-log-logger.md:82]: remove this line

File: doc/plugins/error-log-logger.md
Comment (membphis) [inline doc/plugins/error-log-logger.md:77]: bad title, 'How to set the TCP server address'

File: apisix/cli/ngx_tpl.lua
Comment (membphis) [inline apisix/cli/ngx_tpl.lua:158]: `1m` is small, please change to a larger size.

File: doc/plugins/error-log-logger.md
Comment (dabue) [reply doc/plugins/error-log-logger.md]: updated

File: apisix/cli/ngx_tpl.lua
Comment (dabue) [reply apisix/cli/ngx_tpl.lua]: change to 10m

**Review by membphis — APPROVED** [review]
LGTM

**Comment by membphis** [comment]
@dabue for the doc thing, we need to update the doc index, like this issue: https://github.com/apache/apisix/issues/2790

We can also do this after this PR merge.

File: doc/plugins/error-log-logger.md
Comment (spacewander) [inline doc/plugins/error-log-logger.md:35]: `you're` => your?
`recieve` => `receive`

File: apisix/plugins/error-log-logger.lua
Comment (spacewander) [inline apisix/plugins/error-log-logger.lua:148]: Better to get 3's multiple? Like `get_logs(9)`. And we should reuse `logs` as the second argument of `get_logs`.

File: apisix/plugins/error-log-logger.lua
Comment (spacewander) [inline apisix/plugins/error-log-logger.lua:94]: Style: indent

File: apisix/plugins/error-log-logger.lua
Comment (spacewander) [inline apisix/plugins/error-log-logger.lua:138]: Style: indent

File: doc/plugins/error-log-logger.md
Comment (spacewander) [inline doc/plugins/error-log-logger.md:48]: "chose" => "choose"

File: apisix/plugins/error-log-logger.lua
Comment (spacewander) [inline apisix/plugins/error-log-logger.lua:45]: Need to validate with enum, like this: ` enum = {"GET", "POST", "PUT", "DELETE", "PATCH",`

File: apisix/plugins/error-log-logger.lua
Comment (spacewander) [inline apisix/plugins/error-log-logger.lua:38]: Need to validate with `core.schema.host_def`

**Review by spacewander — CHANGES_REQUESTED** [review]

File: apisix/plugins/error-log-logger.lua
Comment (dabue) [reply apisix/plugins/error-log-logger.lua:148]: LGTM, I'll update it.

File: doc/plugins/error-log-logger.md
Comment (dabue) [reply doc/plugins/error-log-logger.md:35]: :), fix it

File: apisix/plugins/error-log-logger.lua
Comment (dabue) [reply apisix/plugins/error-log-logger.lua:99]: LGTM, I'll fix it.

File: apisix/plugins/error-log-logger.lua
Comment (dabue) [reply apisix/plugins/error-log-logger.lua:38]: that's better.

**Comment by dabue** [comment]
@membphis @spacewander  please review again, thx.

File: .travis/apisix_cli_test.sh.bak
Comment (spacewander) [inline .travis/apisix_cli_test.sh.bak:1]: This file is useless?

File: doc/plugins/error-log-logger.md
Comment (spacewander) [inline doc/plugins/error-log-logger.md:31]: push => pushes

File: doc/plugins/error-log-logger.md
Comment (spacewander) [inline doc/plugins/error-log-logger.md:52]: The `inactive_timeout` in code is 3, but it is 5 in the doc?

File: apisix/plugins/error-log-logger.lua
Comment (spacewander) [inline apisix/plugins/error-log-logger.lua:100]: The name `tls_options` is bad, in fact it isn't related with any options. Would be better to use `tls_sni` or `tls_server_name`.

File: doc/plugins/error-log-logger.md
Comment (spacewander) [inline doc/plugins/error-log-logger.md:50]: The name tls_options is bad

**Review by spacewander — CHANGES_REQUESTED** [review]

File: .travis/apisix_cli_test.sh.bak
Comment (dabue) [reply .travis/apisix_cli_test.sh.bak:1]: It maybe imported by the rebase, I'll clean it.

File: apisix/plugins/error-log-logger.lua
Comment (dabue) [reply apisix/plugins/error-log-logger.lua:100]: got it

**Comment by membphis** [comment]
@dabue you need to rebase your branch

**Comment by dabue** [comment]
@membphis @spacewander  please review again, thx.

**Review by spacewander — APPROVED** [review]

**Comment by membphis** [comment]
@dabue many thx for your nice job

**Pull request merged**
